### ISSUE TITLE
Can't seem to make API calls in onMotionChange callback when app is in background?

### ISSUE THREAD

**Issue by JimmyMultani** [transistorsoft/cordova-background-geolocation-lt#894]
## Your Environment
* Plugin version: 3.0.0-beta.4
* Platform: Android
* OS version: 9.0
* Device manufacturer / model: Xiaomi Mi MIX 2
* Cordova version (`cordova -v`): 8.1.2
* Cordova platform version (`cordova platform ls`): android 7.1.4
* Plugin config
```js
BackgroundGeolocation.ready(
  {
    reset: true, // <-- true to always apply the supplied config
    // Geolocation Config
    desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
    distanceFilter: 0,
    locationUpdateInterval: 5000,
    allowIdenticalLocations: true,
    // Activity Recognition
    stopTimeout: 2,
    activityRecognitionInterval: 5000,
    // Application config
    debug: !debugMode.isRelease, // <-- enable this hear debug sounds.
    logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
    stopOnTerminate: false, // <-- Allow the background-service to continue tracking when app terminated.
    startOnBoot: true, // <-- Auto start tracking when device is powered-up.
    preventSuspend: true, // <-- prevent iOS from suspending the app
    foregroundService: true, // <-- prevent Android from terminating app due to memory pressure
  },
  state => {
    console.log(
      '- BackgroundGeolocation is configured and ready: ',
      state.enabled
    );

    if (!state.enabled) {
      // start tracking
      BackgroundGeolocation.start(function() {
        console.log('- Start success');
      });
    }
  }
);
```

## Expected Behavior

Should make a fetch API call whenever user is switched into `stationary` mode.

## Actual Behavior

Attempts to trigger a fetch call in `onMotionChange` callback but the API response just stalls indefinitely.

## Steps to Reproduce

1. Add this code in addition to the configuration and start the app.
```js
function fetchInfo(data) {
  console.log('checkArea data', data);

  if (data && !data.is_moving) {
    fetch('https://api.github.com/users/transistorsoft/repos').then(
      response => response.json() // this returns a promise
    ).then(repos => {
      console.log('repos',repos);

      LocalNotification.sendNotification({
        title: 'Hello World',
        text: 'Goodbye',
      });
    }).catch(error => {
      console.error(error);
    })
  }
}

BackgroundGeolocation.onMotionChange(motionEvent => {
  const context = `[onMotionChange] Device is ${
    motionEvent.isMoving ? 'moving' : 'still'
  }`;

  console.log(context, motionEvent.location);

  LocalNotification.sendNotification({
    title: context,
    text: JSON.stringify(motionEvent.location),
  });

  fetchInfo(motionEvent.location);
});
```
2. Kill the app.
3. Walk around to trigger the `moving` mode for plugin. The app will start up again.
4. Stop moving so it goes into `stationary` mode.

## Context

I would like to initiate a fetch call to an API when the user goes from `moving -> stationary`, which would in turn execute a local notification.

**Comment by christocracy** [comment]
Install latest [`3.0.0-rc.2`](https://github.com/transistorsoft/cordova-background-geolocation-lt/blob/master/CHANGELOG.md)

Android now has a `#startBackgroundTask` implementation.

```javascript
BackgroundGeolocation.startBackgroundTask((taskId) => {
  LocalNotification.sendNotification({
    title: context,
    text: JSON.stringify(motionEvent.location),
  });
  // Would be nice if LocalNotification provides a callback, then you'd `#stopBackgroundTask`
  // in a callback.
  BackgroundGeolocation.stopBackgroundTask(taskId);
});
```

**Comment by christocracy** [comment]
And see the docs for [`startBackgroundTask`](https://transistorsoft.github.io/cordova-background-geolocation-lt/classes/_cordova_background_geolocation_lt_.backgroundgeolocation.html#startbackgroundtask)

**Comment by JimmyMultani** [comment]
Thanks for the quick response, @christocracy. I will give `startBackgroundTask` a go and report back.

**Comment by christocracy** [comment]
3.0.0-rc.3 is out today with a couple of bug fixes.

**Comment by JimmyMultani** [comment]
Hey @christocracy, I converted `fetchInfo()` to a Promise and implemented `startBackgroundTask` like so:

```js
if (!motionEvent.isMoving) {
  // Perform some long-running task (eg: HTTP request)
  BackgroundGeolocation.startBackgroundTask().then(taskId => {
    fetchInfo(motionEvent.location)
      .then(value => {
        console.log('value', value);
        // When your long-running task is complete, signal completion of taskId.
        BackgroundGeolocation.stopBackgroundTask(taskId);
      })
      .catch(error => {
        // Be sure to catch errors:  never leave you background-task hanging.
        console.error(error);
        BackgroundGeolocation.stopBackgroundTask(taskId);
      });
  });
}
```

Unfortunately, it seems like the API still hangs during the background task. Fetching the API works in all other instances though (in the foreground, in an isolated environment etc.)

I've attached a logcat here: [filtered.log](https://github.com/transistorsoft/cordova-background-geolocation-lt/files/3030155/filtered.log)

I appreciate your assistance. Thank you.

**Comment by JimmyMultani** [comment]
Doesn't seem like fetch API requests can be made through the JS background task when the app is terminated. I had to write it natively and utilize Headless Mode. Using `BackgroundGeolocationHeadlessTask`, `okhttp3`, and Android's `NotificationManager`, I managed to detect when a user is in a stationary state, make a API request to an external URL through an `AsyncTask`, and parse that response data into a notification.

Just leaving this here in case anyone else runs into this same issue.

I'll be looking into iOS now. @christocracy, is there a method similar to Headless Mode that can be used for iOS?

**Comment by christocracy** [comment]
In my Flutter version's sample app, I [test performing an HTTP request in the `onGeofence` callback](https://github.com/transistorsoft/flutter_background_geolocation/blob/master/example/lib/advanced/home_view.dart#L273-L283).  In the response success callback, I play a debug sound `POP`.  This works.  I'm confident this will work with RN too.

>is there a method similar to Headless Mode that can be used for iOS?

iOS has no concept of "running headless".  If the plugin is tracking, your JAvascript app is 100% guaranteed to be alive.

**Comment by christocracy** [comment]
Have you observed the raw `$ adb logcat` of you performing an HTTP request with your callback?

**Comment by JimmyMultani** [comment]
Are you referring to when using `startBackgroundTask()` after the app was terminated? When I was checking the logs at the time, I had a `console.log('searchQuery', searchQuery)` just before the fetch that would list out the URL that the app would initiate a fetch from. The `console.log('response', response)` from the response would never appear though.

I'm probably wrong but it seemed like it had more to do with Chrome not allowing or limiting fetch requests while that app was not running.

**Comment by JimmyMultani** [comment]
You're right, @christocracy. After working a bit on iOS to achieve the same result like on Android, I found that the iOS implementation was much easier. The only issue that I ran into was in regards to CORS issues when using the fetch API, as it would assign the `Origin` as `null`, which Safari is strict about. So if you're using an external API that you have no control over, this can stop you from progressing further. This is not a problem with this plugin intrinsically, and more so something developers will have to consider when working for a similar objective when using this plugin.

Fortunately, a solution that I found for iOS was to use a plugin like [cordova-plugin-advanced-http](https://github.com/silkimen/cordova-plugin-advanced-http) to fetch the data for you in the `startBackgroundTask()`, which will bypass the `Origin: null` issue. (since it's making these fetch requests in a background thread)

Again, just documenting my experience here in case anyone else is trying to achieve a similar objective.
